### DISCUSSION TITLE
VM storage cluster

### DISCUSSION THREAD

**Discussion by ProjectInitiative** [seaweedfs/seaweedfs#3234] [Q&A]
Has anyone successfully deployed seaweedfs as a virtual machine storage pool? I.e. run virtual machine disk image files from `weed mount` or other method?

**Comment by paularlott** [comment]
I have a small proxmox cluster and have mounted seaweedfs via weed mount onto the servers, then added the directory within the proxmox web interface. You can happily deploy a new VM or container onto the mount point and it will run, migration etc will all work.

I did start writing a proxmox plugin to wrap seaweedfs, but I've not finished it yet as earlier versions of seaweedfs failed with qcow2 and large raw files, it has only been the last 2 or 3 versions of seaweedfs that will work correctly with virtual machine images.

I'd say the performance isn't the best for VMs but at the same time I've not really done a lot of testing around using it for VMs, I'm guessing that playing with chunk sizes will yield varying results. I've been trying seaweedfs in various situations to see if I can break it so as to gain confidence in it before deploying it to a production environment.

I use 3 x master, 3 x filers using leveldb2 and 3 or 4 volume servers which are a mix of hdds and ssds. Replication is either 001 or 000, both work fine.

I have also tried using juicefs connected to seaweedfs via the s3 interface, install of the vms is faster but the running of them is slower, again I've only done minimal experiments with this so playing with chunk sizes could yield vastly different results.

Ultimately I would like to be able to use seaweedfs for both file storage within a nomad cluster and also for vm / lxc storage.

**Reply by ProjectInitiative** [reply]
Thanks for the detailed setup! This is almost identical to what I had setup, however my level2db would frequently get out of sync, and vm backups as well as VM disks larger than 2GB would frequently report incorrectly or get corrupted with the mounts. I'll have to try some of the latest versions. This discussion was in the interest of [this issue](https://github.com/chrislusf/seaweedfs/issues/3183) as I'm not quite sure how to provide a reproducible set of instructions of an entire VM cluster.

I've tried lizardfs despite the development being pretty stagnant, and it seems to be more durable against VM corruption, but far too slow to be usable. Ceph was also very slow in comparison to Seaweed, I just had issues with corrupted .raw and .qcow2 files.

**Reply by paularlott** [reply]
I think it was 3.12 where seaweedfs started working with VM files. I was also having issues with out of sync filers but that got resolved in a release a bit before that. I've had 32GB disks on it just fine.

I've been watching that same issue as well and I've tried with your files but it just works and I can't reproduce the checksum error. I've just rebuilt my seaweedfs cluster around the latest version and retesting everything I can think of.

I'm interested in seaweedfs for it's support of large numbers of small files, it handles them a lot better / faster than anything else I've tried.

glusterfs is another option for storing VM files, but not so good with large numbers of small files, however I'd say seaweedfs has the speed and simplicity over glusterfs.

**Reply by ProjectInitiative** [reply]
I had the same issue with 3.12. I wonder if it was due to an already corrupted database. I did try again with TiDB with similar results. I'll have to run some more extensive tests, maybe starting from scratch, I appreciate the info! 

I also really like the small file performance.

**Reply by Stitch10925** [reply]
@paularlott 
What hardware are you using for SeaweedFS? I was aiming to do exactly what you are doing. 

I am currently running SeaweedFS on some LXC's for testing purposes. I am not very confident yet, since I've had some issues with files going missing, but it's probably my own fault. I am fairly new to Linux and didn't setup SeaweedFS properly (I used the default directories which write to /tmp ... which I recently learned get cleared after reboot... unlike in Windows where this is not the case), so I need to start anew and retry.

I have been looking for hardware to run everything on. Initially I was looking at Raspberry Pi (or similar) with a SATA HAT to connect disks to, but I'm worried about the performance.

**Reply by paularlott** [reply]
I'm running it on my test system at present so I have 3 x LXC running on 3 x nucs each LXC has a seaweedfs master, volume server and filer process.

I wouldn't use a PI for a proxmox cluster I have a couple in my cluster but they are limited. Saying that I suspect a PI would work fairly well running just SeaweedFS as it doesn't seem to be that demanding.

**Reply by Stitch10925** [reply]
@paularlott 
Oh no no, I am not going to use PI's for Promox, I have a dedicated machine for that, I just want to use them as storage volumes for SeaweedFS.

Interesting that you run your filer, master and volumes on the same machine. I read about filer sync issues yesterday, what filer storage are you using? SQL?

I was actually thinking of running one machine with all filers and the master, then have different machines with disk-space as volumes. I am still doubting about the filer storage to use, since the filer really needs to be backed up if you do not want to lose your data/file structure. Maybe SQL so it's not on the same machine. This will also allow me to quickly spin up new filers when needed. Maybe on the same machine but with backups? Maybe on the same machine with rsync... not sure yet.

When you get down to the nitty-gritty SeaweedFS is actually quite complex because there are a lot of small things you need to watch out for and take care of. This is what is kind of holding me back because I'm worried I won't be able to recover my data if something goes wrong.

**Reply by paularlott** [reply]
I'm only testing at the moment hence running all 3 parts on the same LXCs, once I'm happy and move to production the plan is 3 x masters in their own containers one per physical server, 2 - 3 x filers in their own containers one per physical server and then each volume server will be in its own container one per server on whichever servers are providing volumes. Then we'll use active-active replication to make some of the data available between data centres.

You want your filers and masters spread around so that if one physical machine fails the majority will still be available to keep the storage available. If you just have 1 machine I'd run just 1 master and 1 filer as there's no point in having more that I can see.

At the moment I'm using leveldb3 for the filers and have a backup process feeding the filer data to MySQL. I've tried using MySQL directly with the filers but leveldb3 is infinitely faster, same goes for CockroachDB etc. I'm undecided as to the exact database we'll use, that'll be the next round of testing once I feel confident that SeaweedFS isn't going to eat my files.

Agreed on the surface it all looks simple but once you dig into it there's a lot of points where it could go wrong, but no worse than any other distributed filesystem and less than some.

The filer sync issue doesn't seem to be an issue any more with the latest version, or not that I've managed to find so far.

**Reply by Stitch10925** [reply]
@paularlott 
Thanks again for the reply!

> You want your filers and masters spread around so that if one physical machine fails the majority will still be available

One issue I have with spreading masters and filers around is the management aspect of it. You need to tell each filer what the IP addresses of the masters are, and you need to tell each volume what the IPs of the masters are. If you are not planning on expanding masters or filers then fine, not much overhead, but especially when expanding (or reducing) masters you need to update all the volume- and filer instances. That seems a bit like a management nightmare to me.

Good point about SQL being slower. That is also why I was contemplating local DB with either rsync or a simple filer-backup. I haven't tried multiple filers yet, but also here I am struggling with the above point: the management aspect of it. If you have multiple filers then you need to make sure all mounts you have contain the IP addresses of all its related filers.

Don't get me wrong. I love that you can create a filer with placement rules or a specific root directory and mount to it. Especially from a security / isolation perspective, but the management of it...

Unlike you, however, I am not looking to set this up with multiple datacenters or whatnot, I am just running a small home-lab.

**Reply by paularlott** [reply]
I use an ansible role to manage my SeaweedFS cluster, it makes life easy if I need to change anything. https://github.com/paularlott/ansible-role-seaweedfs may or may not work for you. I also use Nomad & Consul so between the two systems management is pretty easy.

**Reply by ProjectInitiative** [reply]
I am also using ansible, haven't published anything yet, but the storage plugin looks super interesting

**Reply by paularlott** [reply]
The plugin seems to work but your milage may vary, my perl is somewhat rusty, it's a shame Proxmox doesn't allow you to extend the GUI.

**Reply by ProjectInitiative** [reply]
@paularlott so I rebuilt the cluster from scratch using the latest commit, and I am still unable to run VMs with weed mount. Curious as to why that is the case. What version of proxmox are you using?

**Reply by paularlott** [reply]
Interesting, I still need to rebuild with the last release and retest, but for Proxmox I'm using the latest version 7.2-7.

How big a disk are you using for the VM?
I've used up to 32GB with Ubuntu 22.04 and Debian 11 VMS and 8GB for LXC.

I have my chunk size set to 4MB in the mount, not sure if that makes a huge difference as I've not played with chunk size and performance.

**Reply by ProjectInitiative** [reply]
I am running the same version. I am experiencing the issue with VMs as small as 2GB (low resource linux VM) all the way to 100+GB windows VMs. I am not sure if it is because the disks are Copy-On-Write or not. Will have to keep messing with it.

**Reply by paularlott** [reply]
I'm installing the VMs iSCSI virtual disks and just leaving them as the defaults.

I'm running the 30G version of SeaweedFS as I've only got a small storage space for testing, I've never tried the 800G version. I wonder if it's something to do with that?

**Reply by chrislusf** [reply]
If this can be a reproducible case, e.g., docker compose, it will be easier to debug.

**Reply by ProjectInitiative** [reply]
@chrislusf kinda a hard one to reproduce without a vm cluster, as it is the cluster that is going the reading and writing to the virtual disks

**Comment by chitvan1988** [comment]
Having similar query regarding seaweedfs support of large vm files like qcow2. Are above issues resolved or is much larger disks in the tune of TBs been used/TBs   being tested?

**Reply by chrislusf** [reply]
There are some recent fixes. Please check again with the latest release.
